spec/lib/file-system.spec.js

```diff
@@ -0,0 +1,164 @@
+  EEXIST: 'EEXIST'
+};
+
+describe('The file-system module', () => {
+  let mockfs;
+  let mkdirp;
+  let fs;
+
+  let readDir;
+  let readFile;
+  let writeDir;
+  let writeFile;
+  beforeEach(() => {
+    mockfs = require('mock-fs');
+    mkdirp = require('../../lib/mkdirp');
+    fs = require('../../lib/file-system');
+
+    readDir = 'read/';
+    readFile = {
+      name: 'read.js',
+      content: 'content'
+    };
+    readFile.path = readDir + readFile.name;
+
+    writeDir = 'write/';
+    writeFile = {
+      name: 'write.js',
+      content: 'content'
+    };
+    writeFile.path = writeDir + writeFile.name;
+
+    const config = {};
+    config[readFile.path] = readFile.content;
+
+    mockfs(config);
+  });
+  
+  afterEach(() => {
+    mockfs.restore();
+  });
+
+  describe('The stat() function', () => {
+    it('reads the stats for a directory', done => {
+      fs.stat(readDir).then(stats => {
+        expect(stats).toBeDefined();
+      }).catch(fail).then(done);
+    });
+
+    it('reads the stats for a file', done => {
+      fs.stat(readFile.path).then(stats => {
+        expect(stats).toBeDefined();
+      }).catch(fail).then(done);
+    });
+
+    it('rejects with an ENOENT error on a non-existing directory', done => {
+      fs.stat(writeDir).then(() => {
+        fail('expected promise to be rejected');
+      }).catch(e => {
+        expect(e.code).toBe(ERROR_CODES.ENOENT);
+      }).then(done);
+    });
+
+    it('rejects with an ENOENT error on a non-existing file', done => {
+      fs.stat(writeFile.path).then(() => {
+        fail('expected promise to be rejected');
+      }).catch(e => {
+        expect(e.code).toBe(ERROR_CODES.ENOENT);
+      }).then(done);
+    })
+  });
+
+  describe('The readdir() function', () => {
+    it('reads a directory', done => {
+      fs.readdir(readDir).then(files => {
+        expect(files).toEqual([readFile.name]);
+      }).catch(fail).then(done);
+    });
+
+    it('rejects with ENOENT', done => {
+      fs.readdir(writeDir).then(() => {
+        fail('expected promise to be rejected');
+      }).catch(e => {
+        expect(e.code).toBe(ERROR_CODES.ENOENT);
+      }).then(done);
+    });
+  });
+
+  describe('The mkdir() function', () => {
+    it('makes a directory', done => {
+      fs.mkdir(writeDir)
+        .catch(fail)
+        .then(() => fs.readdir(writeDir))
+        .catch(fail)
+        .then(done);
+    });
+
+    it('rejects with EEXIST', done => {
+      fs.mkdir(readDir)
+        .then(() => fail('expected promise to be rejected'))
+        .catch(e => expect(e.code).toBe(ERROR_CODES.EEXIST))
+        .then(done);
+    });
+  });
+
+  describe('The mkdirp() function', () => {
+    it('makes deep directories', done => {
+      fs.mkdirp(writeDir + readDir).then(() => {
+        return fs.readdir(writeDir + readDir);
+      }).catch(fail).then(done);
+    });
+
+    it('rejects if mkdirp returns an error', done => {
+      pending('figure out how to test this.');
```

Comment:
I was intending to mock mkdirp at first but realized that because the function is imported to be directly accessed, I was unable to do so. Doing only a require('mkdirp'); on the top and calling mkdirp.mkdirp() in the fs.mkdirp function should make this possible.